cylc/flow/xtrigger_mgr.py

```diff
@@ -133,9 +133,13 @@ def __init__(
             TMPL_SUITE_SHARE_DIR: suite_share_dir,
             TMPL_DEBUG_MODE: cylc.flow.flags.debug
         }
-        self.proc_pool: 'SubProcPool' = proc_pool
-        self.broadcast_mgr: 'BroadcastMgr' = broadcast_mgr
-        self.data_store_mgr: 'DataStoreMgr' = data_store_mgr
+
+        if proc_pool is not None:
+            self.proc_pool: SubProcPool = proc_pool
+        if broadcast_mgr is not None:
+            self.broadcast_mgr: BroadcastMgr = broadcast_mgr
+        if data_store_mgr is not None:
+            self.data_store_mgr: DataStoreMgr = data_store_mgr
```

Comment:
Is this to suppress mypy issues when these attrs are used later on? Don't undefined attrs cause mypy failures?

Reply:
1) Yes; 2) No

The only 2 methods that make use of `self.proc_pool`, `self.broadcast_mgr` and `self.data_store_mgr` are `satisfy_xtriggers()` and `callback()`. Perhaps we should instead raise an error at the start of those methods if the required properties are `None`?

Reply:
2. :facepalm:

The `None` pattern here is pretty yuck, just done a quick grep through the code and I think the reason it it like this is because `cylc.flow.config` creates an `XTriggerManager` in validate mode.

If that's the only barrier it should be easy to navigate, `add_trig` does some kind of validation so perhaps:

```python
class XTriggerManager:
   ...
   @staticmethod
   def validate(...):
        ...
    def add_trig(self, ...):
        self.validate(...)
        ...
```

```python
if self.xtrigger_mgr:
    self.xtrigger_mgr.add_trig(...)
else:
    # ensure the validation logic is still carried out by cylc validate
    XtriggerManager.validate(...)
```

Reply:
> protecting you from the little error whilst ignoring the lethal one.

I guess either way it would still fail loudly in the same way?

(I've edited your comment as the indentation was one space off but rendering badly)

Reply:
My idea was to move that into the `validate` method.

Reply:
> I guess either way it would still fail loudly in the same way?

MyPy will fail if the attr is potentially of the wrong type but won't fail if the attr is potentially not set at all.